pkg/datastore/headerscontroller.go

```diff
@@ -0,0 +1,282 @@
+package datastore
+
+import (
+	"sort"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/blocknative/dreamboat/pkg/structs"
+)
+
+type HeaderController struct {
+	content map[uint64]*HNTs
+	ordered []Info
+
+	latestSlot *uint64
+	cl         sync.RWMutex
+}
+
+func NewHeaderController() *HeaderController {
+	latestSlot := uint64(0)
+	return &HeaderController{
+		latestSlot: &latestSlot,
+		content:    make(map[uint64]*HNTs),
+	}
+}
+
+func (hc *HeaderController) CheckForRemoval() (toBeRemoved []uint64, ok bool) {
+	hc.cl.RLock()
+	defer hc.cl.RUnlock()
+	if len(hc.ordered) == 0 {
+		return nil, false
+	}
+
+	l := hc.ordered[len(hc.ordered)-1]
+	for _, v := range hc.ordered {
+		if !(l.Slot-v.Slot > InMemorySlotLag && time.Since(v.Added) > InMemorySlotTimeLag) {
+			return toBeRemoved, ok
+		}
+		toBeRemoved = append(toBeRemoved, v.Slot)
+		ok = true
+	}
+
+	return toBeRemoved, ok
+
+}
+
+func (hc *HeaderController) GetOrderedDesc() (info []uint64) {
+	hc.cl.RLock()
+	defer hc.cl.RUnlock()
+	info = make([]uint64, len(hc.ordered))
+	for i := len(hc.ordered) - 1; i >= 0; i-- {
+		info = append(info, hc.ordered[i].Slot)
+	}
+
+	return info
+}
+
+// addToOrdered to be run in lock
+func (hc *HeaderController) addToOrdered(i Info) {
+	hc.ordered = append(hc.ordered, i)
+	l := len(hc.ordered)
+	if l > 1 {
+		o := hc.ordered[l-1]
+		if o.Slot >= i.Slot {
+			sort.Slice(hc.ordered, func(i, j int) bool {
+				return hc.ordered[i].Slot < hc.ordered[j].Slot
+			})
+		}
+	}
+}
+
+func (hc *HeaderController) GetLatestSlot() (slot uint64) {
+	return atomic.LoadUint64(hc.latestSlot)
+}
+
+func (hc *HeaderController) AddMultiple(slot uint64, hnt []structs.HeaderAndTrace) (err error) {
+	hc.cl.Lock()
+	defer hc.cl.Unlock()
+
+	h, ok := hc.content[slot]
+	if !ok {
+		h = NewHNTs()
+		hc.content[slot] = h
+		hc.addToOrdered(Info{
+			Slot:  slot,
+			Added: time.Now(),
+		})
+
+		ls := atomic.LoadUint64(hc.latestSlot)
+		if ls == 0 || slot > ls {
+			atomic.StoreUint64(hc.latestSlot, slot)
+		}
+	}
+
+	for _, s := range hnt {
+		if err := h.AddContent(s); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (hc *HeaderController) Add(slot uint64, hnt structs.HeaderAndTrace) (newCreated bool, err error) {
+	hc.cl.Lock()
+	defer hc.cl.Unlock()
+
+	h, ok := hc.content[slot]
+	if !ok {
+		h = NewHNTs()
+		hc.content[slot] = h
+		hc.addToOrdered(Info{
+			Slot:  slot,
+			Added: time.Now(),
+		})
+		newCreated = true
+
+		ls := atomic.LoadUint64(hc.latestSlot)
+		if ls == 0 || slot > ls {
+			atomic.StoreUint64(hc.latestSlot, slot)
+		}
+	}
+	return newCreated, h.AddContent(hnt)
+}
+
+func (hc *HeaderController) GetContent(startingSlot, stopSlot uint64, limit int) (elements []structs.HeaderAndTrace, lastSlot uint64) {
```

Comment:
Could we change `content` for `headers`? Here and in the struct name. 

Since the HeaderController, should be working with `headers`, and it's more readable I think.

Reply:
oks added